### ISSUE TITLE
[Request] Downloading while app is closed

### ISSUE THREAD

**Issue by michael-rapp** [haecker-felix/Fragments#30]
Currently the torrents stop downloading/seeding when the Fragments application is closed, aren't they? I suggest to add an option to the settings, which allows to enable downloading/seeding after the application has been closed. This is a common pattern in Gnome for applications that perform background tasks.

Optionally, it might be a good idea to display a permanent notification while the app is running in the background in order to provide the ability to resume it. I think I have read that the Gnome developers suggest to use this pattern to make background operations visible to the user.

**Comment by bertob** [comment]
@michael-rapp no, as far as I know, the current consensus among GNOME designers is that some apps should not run in the background. For some passive things, like messaging apps, this might be desirable once we have a do-not-disturb mode, but for a torrent app I don't think it's a good idea.

**Comment by michael-rapp** [comment]
@bertob As you say, "for some passive things [...] this might be desirable". This is exactly what a Torrent client is to me. I occasionally add some downloads to it, but most of the time the app just runs in background downloading and seeding stuff. I don't need to always keep an eye on it. It is annoying to have the app permanently hanging around while trying to get some other work done. However, I agree that others might use such app differently, just adding a short-running download and closing the window when it's done. For this reason I suggested to make this option optionally.

**Comment by michael-rapp** [comment]
@bertob As you say, "for some passive things [...] this might be desirable". This is exactly what a Torrent client is to me. I occasionally add some downloads to it, but most of the time the app just runs in background downloading and seeding stuff. I don't need to always keep an eye on it. It is annoying to have the app permanently hanging around while trying to get some other work done. However, I agree that others might use such app differently, just adding a short-running download and closing the window when it's done. For this reason I suggested to make this option optionally.

**Comment by michael-rapp** [comment]
I further want to emphasize that my intention was not to use legacy "tray icons" for keeping the app running in background. I looked for official suggestions from Gnome developers for apps that run in background and found the following resources:

* https://wiki.gnome.org/Initiatives/StatusIconMigration/Guidelines
* https://blogs.gnome.org/aday/2017/08/31/status-icons-and-gnome/
* https://wiki.gnome.org/Design/Whiteboards/BackgroundApps

In conclusion, I think that these documents state that it in fact is desirable to keep certain applications running in background, while it is important to provide interaction with such app via notifications.

**Comment by bertob** [comment]
> It is annoying to have the app window hanging around 24/7 while trying to get some other work done.

That's why GNOME has workspaces, so you can put things you don't currently use away, while still keeping them running. Besides, if you're a heavy torrent user you might be better served by an app with a more power-user centric interface anyway.

Most GNOME apps don't run in the background for a reason, and some (e.g. Polari) do it in a way that we're not happy with and that shouldn't be emulated.

Adding preferences because you don't want to decide what kind of user an app is for is a bad way to manage a product. That said, I'd be interested in better understanding the use cases you want to address with this, and if they make sense in our case, we could perhaps find a different solution.

**Comment by michael-rapp** [comment]
> That's why GNOME has workspaces, so you can put things you don't currently use away, while still keeping them running.

Having a workspace just for things that I currently don't want to see seems like a misuse of the workspace concept to me. Such as I don't want to have too many windows open in my main workspace, because it makes switching between windows chaotic, I don't want too many active workspaces, because this also comes with a loss of clarity/tidiness.

> Besides, if you're a heavy torrent user you might be better served by an app with a more power-user centric interface anyway.

I don't identify myself as a power-user as I don't need a very powerful interface with a lot of options. In fact, the simplicity of this app is what I like the most about it. But I do have long running downloads I don't want to keep an eye on all the time. Besides, the whole idea of Torrents relies on people who seed, which is a permanent task that doesn't need any user interaction. I think people are less likely to seed if they are forced to have a window open just for doing so.

> Most GNOME apps don't run in the background for a reason, and some (e.g. Polari) do it in a way that we're not happy with and that shouldn't be emulated.

I agree that many apps misused the ability to run in the background in the past. The concept of having a system tray where background apps lurk around all the time was a really bad idea in the first place and it was a good decision by the Gnome developers to ditch support for it. 

However, there are certain types of apps that run for a very long time without requiring much user interaction while still doing some background work (not just idling). In my opinion for these types of apps it is adequate to allow them to run in background. And I guess a lot of users think the same as many people are sad about the tray option being removed. It might even cause a bad user experience, if an user expects an app to continue its work when closed, but it doesn't. For example, when someone closes his Torrent client expecting that the downloads will proceed, he will be disappointed when returning to the app after hours just to see that no progress has been made. Another example is Gnome Clocks, where alarms stop working when the app is closed. This caused a lot of frustration back when I didn't know about this restriction and missed an important appointment.

I think background apps aren't a bad thing per se, if there is an actual reason for doing so and if the user is always aware of what is going on and can easily resume the app. Therefore it is important to show some kind of indicator to the user. In case of media players this is usually achieved via MPRIS controls. For file synching apps like Dropbox there are plans to add indicators to the Files app. And in case of Gnome Clocks I think it would be a good idea to display a bell icon next to the clock in the top panel just as e.g. Android does. The important thing is that the indicators are shown where the are useful to the user. 

Thinking about this issue more extensively, it possibly isn't the best solution for MPRIS controls to be shown in the notification tray. If there were kind of distinct sound and network menus in Gnome's top panel, like e.g. in Elementary OS, it would be a good idea to show the MPRIS controls in the sound menu and controls for a Torrent app in the network menu.

However, this is currently not technically possible, because of missing APIs and because there are no distinct menus for different topics in nowaday's Gnome. Therefore, in my opinion, the best solution for the Fragments app would be to display a permanent notification, that provides some useful information (e.g. number of uploads/downloads) and allows to quickly pause/resume all activities when running out of bandwidth. 

> Adding preferences because you don't want to decide what kind of user an app is for is a bad way to manage a product.

At least from my experience, most modern Gnome apps that support being run in background have such an option, e.g. Gradio, Lollypop, Polari. And I think it is a good thing to make it opt-in, because nothing (except system processes of course) should run in the background without the user's agreement.

**Comment by bertob** [comment]
> At least from my experience, most modern Gnome apps that support being run in background have such an option, e.g. Gradio, Lollypop, Polari.

That's not true. Of these examples, only Polari is actually part of GNOME, and as I pointed out, we're not happy with its current background behavior. We avoid running things in the background when possible, and aim to have APIs for specific use cases when it is necessary.

Solutions that leave it up to developers whether their app should run in the background and show some kind of indicator are going to get abused one way or another. It's happened with status icons on the desktop and is now happening again with persistent notifications on Android.

> Therefore, in my opinion, the best solution for the Fragments app would be to display a permanent notification, that provides some useful information (e.g. number of uploads/downloads) and allows to quickly pause/resume all activities when running out of bandwidth.

No, hacking notifications that way is a bad idea. Not only because it's not a great fit for this particular problem, but also because it sets a bad precedent for the rest of the ecosystem.

If you want a way for torrent apps to run in the background, propose an API upstream that provides a solution for the specific use cases you're interested in (a good starting point might be discussing this in #gnome-design or #gnome-hackers on IRC). Until we have that, I think we should close this issue.

I realize that this might be frustrating to you, but this is the way GNOME works. We don't do hacks. We fix the root problem, even if it takes a long time.

**Comment by michael-rapp** [comment]
I think we reached a point, where we don't exclusively talk about the Fragments app anymore, but discuss the abilities, restrictions and ideals of Gnome, which is beyond the scope of this feature request. @bertob, allow me to summarize our arguments so far. If you feel that I cited one of your points inaccurately, please correct me.

* You were against the ability to run a torrent client in the background in the first place. I argued that, depending on the use case, such app might perform very long running tasks and therefore should be able to do so. I think, by giving examples, I emphasized that the discussed feature is actually desirable by some and I feel that this requirement is actually compliant with Gnome's guidelines.
* You were against making the discussed feature opt-in by providing the ability to enable it in the app's settings. Your argument against this was that it would be bad to not decide what kind of users the app targets. I think that the Fragments app is already well-suited for users that occasionally download small files as well as for those that download many large files and seed regularly. In my opinion, adding such an option would enable the app to be flexibly adapted to the needs of the latter while not affecting the default experience and therefore I feel like this would be an improvement over all.
* If the ability to run the app in background was added, you would be against displaying a permanent notification to make its state visible to the user. You consider this a workaround and are afraid that other third-party apps will misuse this concept such as it happened with status icons. I agreed that such a solution is not ideal but the only one that is currently feasible due to missing APIs. Personally, I would also be fine with not providing any information about the background process to the user, but I think that this is not ideal either. We both agree, that on the long run, the best solution would probably be to propose an API for propagating long-running background network activities to the user such as [libcloudproviders](https://gitlab.gnome.org/World/libcloudproviders) does for file sync clients. However, a good approach how (and where) this could be integrated in Gnome's UI is still missing and it is questionable if such an effort would be accepted by the Gnome developers.

From my point of view, we could proceed in one of the following ways:

* Decide not to implement the ability to run Fragments in the background at all.
* Decide not to implement the ability to run Fragments in the background as long as there is no specialized integration API for background network activities. Are we able to propose such API?
* Decide to implement the ability to run Fragments in the background. This requires to make a decision for each one of the following issues:

1.  Should a permanent notification be displayed when running in the background? Or should notifications only be shown, when a task (e.g. a download) has finished (cf. https://github.com/haecker-felix/Fragments/issues/31)
2.  Should the feature be enabled by default?
3.  Should there be an option to disable/enable this behavior?

> I think we should close this issue.

I think we should leave it up to @haecker-felix to close this issue or to keep it open based on the arguments and counter-arguments we brought up.

**Comment by bertob** [comment]
@michael-rapp sorry for the slow reply. I think your summary is accurate, my basic position is
- I don't think keeping a window around is a big problem
- If we were to run the app in the background we should properly think through the experience and do it right, rather than adding random settings
- Doing that would probably require new APIs which GNOME does not have at the moment

As I said, the GNOME way to solve this problem is doing the necessary design work, discussing it upstream, and then proposing and implementing new APIs that cover our use cases. This means thinking about all the potential pitfalls and edge cases (what if we're on a metered connection? What if someone lives in a country that punishes torrenting? What if they need the bandwidth for something else? etc.), and coming up with a solution that mitigates these issues in a clean way. Once all of that is addressed we can safely enable the behavior by default, and probably wouldn't even need a preference for it.

We're definitely able to propose such an API, it just requires someone to sit down and do the work ;)

The reason why I'm personally not particularly interested in this at the moment is that it's a lot of work for relatively little gain. That said, if you or someone else wanted to work on it that would be awesome and I'd be happy to support the effort!

**Comment by haecker-felix** [comment]
I'll close this issue, since the missing API is right now not a issue of Fragments itself. As soon as such API exists, I will implement it.

Thanks for this healthy, constructive discussion.
